content/blog/2025-12-16-facade-layout-compiler.md

```diff
@@ -0,0 +1,409 @@
+
+#### Window Spacing Rule
+
+If the rule `min_window_spacing` is declared, the analyzer checks horizontal distances between adjacent windows in each row. Unlike door and chimney rules, spacing violations produce warnings only rather than errors, because tight window spacing may be intentional in some architectural styles.
+
+#### Symmetry Detection
+
+The analyzer compares each row with its horizontal mirror. Asymmetric rows produce warnings to inform the designer of potential stylistic inconsistency. This is informational rather than prescriptive.
+
+### Error Classification
+
+I deliberately distinguished between errors (fatal violations) and warnings (stylistic concerns):
+
+- **Error:** Structurally invalid facade. Compilation halts; no output generated.
+- **Warning:** Stylistic issue or optional constraint violation. Compilation continues; warning recorded in output.
+- **Info:** Informational message (e.g., auto-fill notification). No effect on compilation.
+
+This classification prevents the compiler from being overly restrictive while still catching genuine architectural errors.
+
+---
+
+## Back-End: Code Generation
+
+For valid programs, the compiler generates two output formats:
+
+### JSON Output
+
+The JSON output includes:
+
+- Grid metadata (dimensions, rules, auto-filled cells)
+- The normalized grid as a 2D array of symbols
+- Validation results (is_valid, errors, warnings)
+
+This format is designed for programmatic consumption. The structured data can be used by downstream tools for further processing, constraint solving, or generative design.
+
+### SVG Output
+
+The SVG generator produces a visual representation where each cell is rendered as a colored rectangle with a label. Cells are grouped by type in the SVG DOM, making it easy to style or manipulate specific element types.
+
+### PNG Rasterization
+
+For environments that don't support SVG, the compiler can produce PNG output via CairoSVG.
+
+---
+
+## What were the hardest parts to get right?
+
+### 1. Separating Syntax Errors from Semantic Errors
+
+The distinction between lexical, syntactic, and semantic errors was initially blurry. For example:
+
+- Is `@` an invalid character (lexical) or an invalid symbol (semantic)?
+- Is `row 1: X E E` a syntax error or a semantic error?
+
+I resolved this by establishing boundaries:
+
+- **Lexical errors:** Characters not in the DSL's alphabet (punctuation, Unicode, emoji)
+- **Syntax errors:** Malformed constructs (missing colons, invalid repeat expressions, missing row numbers)
+- **Semantic errors:** Valid syntax with invalid meaning (floating doors, disconnected chimneys, invalid symbol letters like `X`)
+
+The key insight was that `X` is a valid identifier syntactically but not a valid symbol semantically. The parser accepts it and emits a semantic error during cell processing.
```

Comment:
This seems like a long way to say: "errors can be emitted at different parts of the compiler." I'm not sure what the insight is here. It's also not clear *why* you decided that `X` should be "a valid identifier syntactically but not a valid symbol semantically"; other choices also seem like they could work.